glusterd2/pmap/registry.go

```diff
@@ -0,0 +1,207 @@
+package pmap
+
+import (
+	"expvar"
+	"fmt"
+	"net"
+	"sync"
+)
+
+const (
+	gfIanaPrivPortsStart = 49152
+	gfPortMax            = 65535
+)
+
+// PortState represents the type of state of the port
+type PortState int32
+
+// List of states an individual port can be in
+const (
+	GfPmapPortFree PortState = iota
+	GfPmapPortForeign
+	GfPmapPortLeased
+	GfPmapPortInUse
+)
+
+type portStatus struct {
+	Port  int       `json:"port"`
+	State PortState `json:"state"`
+	Conn  net.Conn  `json:"-"`
```

Comment:
A suggestion here. Let me know what you think. We could have `portStatus` also track the bricks status. And in place of the brickpath->brickStatus map, we have a brickpath->port and conn->port maps. This would allow quick lookups later on, avoiding iteration, especially when dealing with a large number of bricks.

So we'd have,

```go
type portStatus struct {
// Don't need the Port number in the status. This will be in the correctly indexed position in the ports slice.
  State PortState
  Conn net.Conn
  Bricks map[string]bool
}

type pmapRegistry struct {
  .
  .
  bricks map[string]int
  conns map[net.Conn]int
 .
 .
}
```

Having a conn->port map will make it much simpler to delete whole connections now.
```go
func (r *PmapRegistry) RemoveByConn (conn net.Conn) error {
  p := r.conns[conn]

  delete(r.conns, conn)

  for brick, _ := range p.Bricks {
    delete(r.bricks, brick)
  }
  p.Status = PortFree
  p.Conn = nil
  p.Bricks = make(map[string]bool)
}
```

Reply:
I did think about `conn->port` mapping earlier and removing `Port` field. Had decided to do that later. Now that you've pointed it out, I'll update it in the same PR :+1:

Reply:
Done.

Removed `Conn` field from portStatus as it wasn't being used anywhere.